internal/webhooks/machineset.go

```diff
@@ -219,3 +253,104 @@ func validateSkippedMachineSetPreflightChecks(o client.Object) *field.Error {
 	}
 	return nil
 }
+
+// calculateMachineSetReplicas calculates the default value of the replicas field.
+// The value will be calculated based on the following logic:
+// * if replicas is already set on newMS, keep the current value
+// * if the autoscaler min size and max size annotations are set:
+//   - if it's a new MachineSet, use min size
+//   - if the replicas field of the old MachineSet is < min size, use min size
+//   - if the replicas field of the old MachineSet is > max size, use max size
+//   - if the replicas field of the old MachineSet is in the (min size, max size) range, keep the value from the oldMS
+//
+// * otherwise use 1
+//
+// The goal of this logic is to provide a smoother UX for clusters using the Kubernetes autoscaler.
+// Note: Autoscaler only takes over control of the replicas field if the replicas value is in the (min size, max size) range.
+//
+// We are supporting the following use cases:
+// * A new MS is created and replicas should be managed by the autoscaler
+//   - Either via the default annotation or via the min size and max size annotations the replicas field
+//     is defaulted to a value which is within the (min size, max size) range so the autoscaler can take control.
+//
+// * An existing MS which initially wasn't controlled by the autoscaler should be later controlled by the autoscaler
+//   - To adopt an existing MS users can use the default, min size and max size annotations to enable the autoscaler
+//     and to ensure the replicas field is within the (min size, max size) range. Without the annotations handing over
+//     control to the autoscaler by unsetting the replicas field would lead to the field being set to 1. This is very
+//     disruptive for existing Machines and if 1 is outside the (min size, max size) range the autoscaler won't take
+//     control.
+//
+// Notes:
+//   - While the min size and max size annotations of the autoscaler provide the best UX, other autoscalers can use the
+//     DefaultReplicasAnnotation if they have similar use cases.
+func calculateMachineSetReplicas(ctx context.Context, oldMS *clusterv1.MachineSet, newMS *clusterv1.MachineSet, dryRun bool) (int32, error) {
```

Comment:
have we considered passing oldReplicas, newReplicas through the signature of this function instead of the  machineSet resources and reused it for both MD and MS? It seems the objects are only needed for if oldMD == nil { which would actually be the same code path than case oldMD.Spec.Replicas == nil:

Reply:
If we just use one generic function instead of duplicating I would prefer to pass in oldObject (or oldMDorMS) of type runtime.Object instead of merging the two code paths of oldMS == nil & oldMS.Spec.Replicas in the func

Reply:
ok cool, let's keep current implementation since it seems to have general lgtm and this code is unlikely to change often.